File: 1425-constrained-subsequence-sum/1425-constrained-subsequence-sum.py
class Solution:
    def constrainedSubsetSum(self, nums: List[int], k: int) -> int:
        dq = deque()
        for i in range(len(nums)):
            nums[i] += dq[0] if dq else 0
            while dq and dq[-1] < nums[i]:
                dq.pop()
            if nums[i] > 0:
                dq.append(nums[i])
            if i >= k and dq and nums[i - k] == dq[0]:
                dq.popleft()
        return max(nums)

File: 1425-constrained-subsequence-sum/1425-constrained-subsequence-sum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var constrainedSubsetSum = function(nums, k) {
  const queue = [];
  let [l, r, ans] = [0, -1, nums[0]];
  
  for (let i = 0; i < nums.length; i++) {
    nums[i] += (l <= r ? nums[queue[l]] : 0);
    ans = Math.max(ans, nums[i]);
    while (l <= r && nums[queue[r]] <= nums[i]) r--;
    
    if (nums[i] > 0) queue[++r] = i;
    if (l <= r && i - queue[l] >= k) l++;
  }
  
  return ans;
};